Support merging

hansSchall opened the pull request:
fixes #2

hansSchall commented:
This overlay for `Receiver` will store the latest values for all sources for the specified time.
When any source changes, the merged values will be recomputed and changes within the output will be delivered to the user, by using the `changed` event.

Usage:
````ts
const rec = new ReceiverMerge({
    universes: [1],
    reuseAddr: true,
    timeout: 5000, // time the values will be stored before they are droped (assuming the source is down / has stopped sending data)
    // ... more options (everything here will be passed to new Receiver({...}))
})

rec.on("changed", ({ addr, newValue: value }) => { console.log(addr, value) });
````

k-yle commented on src/merge.ts:
```diff
@@ -0,0 +1,102 @@
+}
+
+export class ReceiverMerge extends Receiver {
+    constructor({
+        universes = [1],
+        port = 5568,
+        iface = undefined,
+        reuseAddr = false,
+        timeout = 5000,
+    }: MergeProps) {
+        super({
+            universes,
+            port,
+            iface,
+            reuseAddr
+        });
```
```suggestion
    constructor({ timeout = 5000, ...props }: MergeProps) {
        super(props);
```

k-yle reviewed (changes requested):
This looks great, thanks! I'll have some time next week to properly test it.

hansSchall commented on src/merge.ts:
```diff
@@ -0,0 +1,102 @@
+        let i = 0;
+        while (i < 512) {
+            if (this.lastData.data[i] != mergedData.data[i]) {
+                super.emit("changed", {
```
The other changes @k-yle souggested are quite usefull, but I am not a friend of this change.

Most users (99.9%) would do something like this:

````js
rec.on("changed", changes => {
  for(let change of changes){
    // do exactly the same as before
  }
})
````

In the few use cases, where caching is needed, you should use your own caching algorithm, to cache data from different universes and sources, too. This would be more efficient.

The only reason to implement it like this could be the Node.js native EventEmitter's performance; but the word *performance* doesn't appear in the official documentation.
If you think, that this point is important, please change it.

hansSchall replied:
I can understand your thought, that too many emitted events could affect the performance negatively.
A simple solution could be to move the `EventEmitter.emit()` call into a `process.nextTick`.
This way the event would be emitted asynchronously.

To improve external caching, we could emit a second event which indicates, that the full universe has been merged.

What do you think about this?

hansSchall commented on src/merge.ts:
```diff
@@ -0,0 +1,102 @@
+    mergePacket(packet: Packet) {
+        // used to identify each source (cid & universe)
+        let pid: string = packet.cid.toString() + "#" + packet.universe.toString();
+        this.senders.set(
```
I've added both events. They could be quite usefull.


In the future we could create a more detailed event object and a sender list.

hansSchall commented:
I have coded more detailed event objects; now they contain the senders cid and universe

### API
````ts
merger.on("senderConnect", (ev: {cid: string, universe: number, firstPacket: Packet}) => {
  ...
});
merger.on("senderDisconnect", (ev: {cid: string, universe: number, lastPacket: Packet}) => {
  ...
});
````

additionally I've added default values to comply with ts-strict and converted SendersData to `interface SendersData`

mutec commented:
Hi there, it would be cool to support switching between LTP and HTP.
Are there any plans to publish the merge-feature in the near future?

k-yle commented:
Thanks again for making this PR. My original concerns from https://github.com/k-yle/sACN/pull/43#discussion_r845638359 still apply, but firstly: the performance of nodejs v21's EventEmitter is much better than node v13, and secondly: I'd rather merge this PR than have it sit around for any longer.
